### PR TITLE
63 pluto corrections applied and not applied

### PR DESCRIPTION
**Pull request by abrieff**
There are alot of changes in this PR that are just moving the existing pluto versions reports into a function to facilitate a seperate page for the Manual Corrections. The additions to pluto.py are at the bottom in the manual_corrections_report function.

This PR is blocked by getting the manual corrections applied and not applied reports pushed from PLUTO.

1. New Page for Manual Corrections shows graphs by field for corrections applied and not applied.
Includes Dropdown to filter by Version or all Records.
2. Incorporates Language from our existing readme for pluto corrections
3. Follows FACDB architecture for two different pages on the same app - I wonder if we don't want to decouple these even more in the future and create new seperate component files. I find all of the nested function stuff hard to follow, but I tried to structure this code so that they could easily be refactored in the future.

### CODE REVIEW

**Comment by abrieff** [comment]
Screen Shot 2022-07-19 at 4 17 05 PM
Screen Shot 2022-07-19 at 4 17 29 PM
Screen Shot 2022-07-19 at 4 17 36 PM

**Comment by SashaWeinstein** [comment]
The refactor of `pluto.py` makes it hard for me to understand the changelog. I agree that `pluto.py` needs a refactor, but I don't know if that change should be combined with this PR. We should discuss that refactor together so that I can understand what I'm seeing

**Comment by SashaWeinstein** [comment]
Ok nevermind there weren't actually changes to the code made. Andrew do you know why git think almost the entire file was changed?

**Comment by SashaWeinstein** [comment]
Seems like [this commit to create a new section](https://github.com/NYCPlanning/data-engineering-qaqc/pull/70/commits/f8f94a7e54129409e21f8c38d7ff625c41ab82ad) made a ton of changes that I don't understand

**Comment by abrieff** [comment]
Sorry Sasha - mentioned it at the top but there are no changes to the existing code other than they are put inside of a new version_comparison_report function. I will add comments to highlight where there are actually code changes

File: src/pluto/pluto.py [hunk 1]
```diff
@@ -421,111 +321,319 @@ def generate_graph(v1, v2):
...
-        ]:
-            val1 = [i["values"] for i in v1_exp if i["field"] == field][0]
-            val2 = [i["values"] for i in v2_exp if i["field"] == field][0]
-            in1not2 = [i for i in val1 if i not in val2]
-            in2not1 = [i for i in val2 if i not in val1]
-            if len(in1not2) == 0 and len(in2not1) == 0:
-                pass
+        create_expected(data["df_expected"], v1, v2)
+
+    def manual_corrections_report(data):
```
Comment (abrieff) [inline src/pluto/pluto.py:535]: New code starts here

File: src/pluto/pluto.py [hunk 2]
```diff
@@ -421,111 +321,319 @@ def generate_graph(v1, v2):
...
+        create_corrections_section(not_applied_corrections, version, applied=False)
+
+        st.info(
+            """
+            See [here](https://www1.nyc.gov/site/planning/data-maps/open-data/dwn-pluto-mappluto.page) for a full accounting of the changes made for the latest version
+            in the PLUTO change file.
+            """
+        )
+
+    report_type = st.sidebar.selectbox("Choose a Report Type", ["Compare with Previous Version", "Review Manual Corrections"])
```
Comment (abrieff) [inline src/pluto/pluto.py:633]: This is the logic for the different pages of PLUTO - all existing pluto logic (unchanged) is in the version_comparison_report function below.

**Comment by SashaWeinstein** [comment]
Ok word. No need to apologize, I just rely on those green/red highlights so much haha. Can you explain more about why moving "existing pluto versions reports into a function" helps to "to facilitate a seperate page for the Manual Corrections"? I think understanding the context would help me see how these changes all belong together in one PR instead of two separate ones

**Comment by SashaWeinstein** [comment]
Ok perfect! Thanks for spelling this out for me. I agree that an if/else block is better here. I have one comment on the code, which is to have an `elif` with the other dropdown option, as it makes the code more readable and extendable. 

Can you write up an issue articulating why pluto should be split up into "component files", and include your understanding of "components" in this context? I like that language a lot, that issue could be instructive on how we divide data product pages into modules. Which we definitely should do

File: src/pluto/pluto.py (hunk 2 above)
Comment (SashaWeinstein) [reply src/pluto/pluto.py:634]: ok word, thanks for helping me navigate this. My first comment before I dig into the new content is that this should be moved up to live below the `get_data` call. My view of how `pluto.py` should be structured:
 1. interactive elements that let user chose what content they want to see
 2. function definitions
 3. calls to those function 

Does this seem like a good plan to you?

File: src/pluto/pluto.py [hunk 3]
```diff
@@ -421,111 +322,319 @@ def generate_graph(v1, v2):
...
+
+
+        st.header("Manual Corrections")
+
+        applied_corrections = data['pluto_corrections_applied']
+        not_applied_corrections = data['pluto_corrections_not_applied']
+
+        if applied_corrections is None or not_applied_corrections is None:
+            st.info("There are no available corrections reports for this branch.")
+            return
```
Comment (SashaWeinstein) [inline src/pluto/pluto.py:594]: This block doesn't seem necessary to me. Doesn't it make more sense to get the associated table from `get_data` inside the `create_corrections_section` function so we only have to pass one variable? 
Also this approach would lead not visualizing corrections applied if corrections not applied was none and vice versa. 
Also I think we want to check the `.empty` method of the dataframes  instead of whether they are `None`. This check is good, it should just happen in the `create_corrections_section` call

File: src/pluto/pluto.py (hunk 2 above)
Comment (abrieff) [reply src/pluto/pluto.py:634]: Yeah that seems reasonable!

File: src/pluto/pluto.py [hunk 4]
```diff
@@ -421,111 +322,319 @@ def generate_graph(v1, v2):
...
+
+        st.subheader("Manual Corrections Not Applied", anchor="corrections-not-applied")
+
+        st.markdown(
+            """ 
+            For each record in the PLUTO Corrections table, PLUTO attempts to correct a record by matching on the BBL and the 
+            Old Value column. As the underlying datasources change and improve, PLUTO records may no longer match the old value 
+            specified in the pluto corrections table. The graph and table below outline the records in the pluto corrections table that failed to be applied for this reason.
+            """
+        )
```
Comment (SashaWeinstein) [inline src/pluto/pluto.py:626]: Can these elements also go inside the `create_corrections_section`? Conceptually to me they are part of the "corrections section" as a part of the app. 
Maybe a dictionary that maps the text of "applied" and "not applied" to the correct subheader/markdown elements? Maybe that's overkill and all we need is and if/else block. Curious about your thoughts

**Review by SashaWeinstein — COMMENTED** [review]
Love the aggrid, great upgrade. I have two comments on the front-end I want to share:
1. I think having a block of text from a markdown element for the correction applied between the subheader and the graph similar to the corrections not applied would be nice. I think it can be much shorter, something like "corrections are applied only when the existing value exactly matches the corrections table". More for symmetry than anything else. Maybe this is just OCD and we can leave it
2. I imagined that we would keep in the table of all corrections and from there break it down into applied and not applied. Not sure if this would be better, and curious if you thought about this and decided against it or if there was other conversation I missed.

**Comment by abrieff** [comment]
@SashaWeinstein 
1. Sure, I can add something.
2. No conversation you missed, I just figured it would end up being more confusing than it's worth, since it's not really telling you anything about the changes actually made to PLUTO. I think in particular, having it be the first thing on the report you see would be confusing as it would draw your attention to something that conveys less information than the other two charts. But I don't have as great of a view into the use-case, so I could be wrong.

File: src/pluto/pluto.py (hunk 4 above)
Comment (abrieff) [reply src/pluto/pluto.py:626]: I think I disagree here, but its maybe because create_corrections_section is misnamed. It's a function for wrapping the logic around displaying a graph and underlying dataframe. If we wanted to encapsulate this logic in functions, i'd suggest seperate wrapper functions around create_corrections_section instead like
```
def corrections_applied_section():
  # markdown
  # display_corrections_graphs
  # info
```

File: src/pluto/pluto.py (hunk 3 above)
Comment (abrieff) [reply src/pluto/pluto.py:594]: This is testing something slightly different than the checks that exist in the create_corrections_section for whether there are no corrections for the version - it's for if the branch itself has no files for the reports in Digital Ocean. I think it might be unnecessary since it's mostly testing for a condition most likely to occur to a developer, but there's no real time at which this 
```
Also this approach would lead not visualizing corrections applied if corrections not applied was none and vice versa.
```
would be relevant, since the files are most likely both there or both not there. 

 I could alternatively catch the no files thing and return an empty dataframe instead of None (see https://github.com/NYCPlanning/data-engineering-qaqc/pull/70/files#diff-57a9a9900497ce261b5bb70460786b26bbc676bca5270cf37efd9da08368de96R108) and let the checks in create_corrections_section handle the nothing to display issue, but that will tell the end-user less about what the actual problem is

File: src/pluto/pluto.py [hunk 5]
```diff
@@ -421,111 +322,319 @@ def generate_graph(v1, v2):
...
+
+        create_expected(data["df_expected"], v1, v2)
+
+    def manual_corrections_report(data):
+        def create_corrections_section(df, version, applied):
+            def title_text(version, applied):
+                applied_text = "Applied Manual Corrections " if applied else "Manual Corrections Not Applied "
+                version_text = "for All Versions by Field" if version == "All" else f"introduced in Version {version} by Field"
+
+                return applied_text + version_text
```
[the remark this answers is not in the document]
Comment (abrieff) [reply src/pluto/pluto.py:544]: 👍

File: src/pluto/pluto.py (hunk 4 above)
Comment (abrieff) [reply src/pluto/pluto.py:626]: On second thought, I like the dictionary idea I think - will try to make this cleaner

File: src/pluto/pluto.py (hunk 3 above)
Comment (SashaWeinstein) [reply src/pluto/pluto.py:594]: Ok great! That makes sense. Although I think giving a dictionary a non-existent key would prompt an error instead of returning `None`. The dictionary's `.get` method would return `None` if no such key exists.

I still like passing `create_corrections_section` one variable for whether it's corrections applied or not applied, and setting everything based on that one variable. I know this is quite lightweight but it would make the code more readable

**Comment by SashaWeinstein** [comment]
Ok wonderful, that makes sense to not have the total list of possible corrections in one graph.

Also I think the merge conflict is from me adding the new version to the list, I can resolve this if you want

File: src/pluto/pluto.py [hunk 6]
```diff
@@ -422,111 +325,325 @@ def generate_graph(v1, v2):
...
-    # EXPECTED VALUE
-    st.header("Expected Value Comparison")
-    st.write(
-        "if nothing showed up, then it means there aren't any expected value change"
-    )
+        applied_corrections = data['pluto_corrections_applied']
+        not_applied_corrections = data['pluto_corrections_not_applied']
+
+        if applied_corrections is None or not_applied_corrections is None:
+            st.info("There are no available corrections reports for this branch.")
```
Comment (SashaWeinstein) [inline src/pluto/pluto.py:629]: Maybe we could include a sentence like "this is probably due to a problem on the backend" or something like that?

Comment (SashaWeinstein) [inline src/pluto/pluto.py:629]: Maybe we could include a sentence like "this is probably due to a problem on the backend" or something like that?

**Review by SashaWeinstein — APPROVED** [review]
This looks good to merge! I thought having one `create_corrections_graph` was nicer than the two `applied_corrections_section` and `not_applied_corrections_section` but I think it's fine as is
